File: backend/japanese/src/main/java/io/github/tuanpq/japanese/grammar/GrammarService.java
package io.github.tuanpq.japanese.grammar;

import java.util.List;
import java.util.Optional;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;

@Service
public class GrammarService {

    private final GrammarRepository grammarRepository;

    public GrammarService(GrammarRepository grammarRepository) {
        this.grammarRepository = grammarRepository;
    }

    public List<Grammar> findAllGrammars() {
        return grammarRepository.findAll();
    }

    public Optional<Grammar> findGrammarById(Long id) {
        return grammarRepository.findById(id);
    }

    public Page<Grammar> findGrammars(int page, int size) {
        return grammarRepository.findAll(PageRequest.of(page, size));
    }

    public Grammar addGrammar(Grammar grammar) {
        return grammarRepository.save(grammar);
    }

    public Grammar updateGrammar(Long id, Grammar grammar) {
        grammar.setId(id);
        return grammarRepository.save(grammar);
    }

    public void deleteGrammar(Long id) {
        grammarRepository.deleteById(id);
    }

    public void deleteGrammarsByIds(List<Long> ids) {
        grammarRepository.deleteAllById(ids);
    }

}

File: mobile/android/app/src/main/java/io/github/tuanpq/data/remote/GrammarService.kt
package io.github.tuanpq.data.remote

import io.github.tuanpq.data.remote.model.Grammar
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path

interface GrammarService {

    @GET(API.FIND_ALL_GRAMMARS)
    suspend fun findAllGrammars(): Response<List<Grammar>>

    @GET(API.FIND_GRAMMAR_BY_ID)
    suspend fun findGrammarById(@Path("id") id: Int): Response<Grammar>

    @POST(API.ADD_GRAMMAR)
    suspend fun addGrammar(@Body grammar: Grammar): Response<Grammar>

    @PUT(API.UPDATE_GRAMMAR)
    suspend fun updateGrammar(@Path("id") id: Int, @Body grammar: Grammar): Response<Grammar>

    @DELETE(API.DELETE_GRAMMAR)
    suspend fun deleteGrammar(@Path("id") id: Int): Response<Void>

    @DELETE(API.DELETE_GRAMMAR_BY_IDS)
    suspend fun deleteAllGrammarsByIds(@Body ids: List<Int>): Response<Void>

}

